### PR TITLE
CIL callback

### PR DESCRIPTION
**Pull request by paskino**
Adapt as a CIL `Callback`

### CODE REVIEW

File: img_quality_cil_stir/image_quality_callback.py [hunk 1]
```diff
@@ -93,15 +125,30 @@ def __init__(self, reference_image,
...
             NotImplementedError
             
-    def eval(self, iteration, last_cost , test_image):
+    def __call__(self, algorithm):
+        iteration = algorithm.iteration
+        last_cost = algorithm.get_last_objective(return_all=False)
+        # evaluate the test image
+        self.eval(iteration, last_cost, algorithm.x)
+        
+    def eval(self, iteration, last_cost, test_image):
```
Comment (casperdcl) [inline img_quality_cil_stir/image_quality_callback.py]:
```suggestion
        test_image = algorithm.x
```

File: examples/recon_example.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+#%%
+show2D([ground_truth, top, bottom], num_cols=3)
+#%%
+# setup a 2 binary ROI images
+roi_image_dict = {
+    'top' : top,
+    'bottom' : bottom
+}
+
+from cil.utilities.quality_measures import mae, mse, psnr
```
Comment (casperdcl) [inline examples/recon_example.py]: unused?

File: examples/recon_example.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+def MSE(x,y):
+    """ mean squared error between two numpy arrays
+    """
+    return ((x-y)**2).mean()
```
Comment (casperdcl) [inline examples/recon_example.py]: MSE is generally a horrible metric as it gives too much weight to background.
Would strongly recommend NRMSE instead:

```suggestion
def NRMSE(x, y):
    """normalised mean squared error between two arrays"""
    return (((x-y)**2).mean() / (y**2).mean()) ** 0.5
```

File: examples/recon_example.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        y is considered to be the reference array and the default scale
+        needed for the PSNR is assumed to be the max of this array
+    """
+  
+    mse = ((x-y)**2).mean()
+  
+    if scale == None:
+        scale = y.max()
+  
+    return 10*np.log10((scale**2) / mse)
```
Comment (casperdcl) [inline examples/recon_example.py:102]: would drop PSNR as it's extremely sensitive to noise.

File: examples/recon_example.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+def MSE(x,y):
+    """ mean squared error between two numpy arrays
+    """
+    return ((x-y)**2).mean()
+
+def MAE(x,y):
+    """ mean absolute error between two numpy arrays
+    """
+    return np.abs(x-y).mean()
```
Comment (casperdcl) [inline examples/recon_example.py:89]: would drop this

**Review by casperdcl — CHANGES_REQUESTED** [review]
In addition to NRMSE, I'd recommend using Mean structural similarity (MSSIM[^1]). Let me know if you want me to add an implementation of it :)

MSE/MAE/PSNR are appallingly poor metrics in most scenarios IMO.

[^1]: Wang, Bovik, Sheikh, and Simoncelli (2004) IEEE TIP 13(4) 600-612, "Image Quality Assessment: From Error Visibility to Structural Similarity"

File: examples/recon_example.py (hunk 3 above)
Comment (KrisThielemans) [reply examples/recon_example.py]: we can have both. For PETRIC, we want to use RMSE over (targeted) VOI (normalised by mean of background VOI

File: examples/recon_example.py (hunk 4 above)
Comment (KrisThielemans) [reply examples/recon_example.py:102]: sure, but many people will want it (but we don't)

File: examples/recon_example.py (hunk 5 above)
Comment (KrisThielemans) [reply examples/recon_example.py:89]: as above

**Comment by KrisThielemans** [comment]
> In addition to NRMSE, I'd recommend using Mean structural similarity (MSSIM[1](#user-content-fn-1-a07db8b5583dc14a747b01d4a6fb9512)). Let me know if you want me to add an implementation of it :)

I'm sure people will appreciate that, but let's keep that for later 😄 

> MSE/MAE/PSNR are appallingly poor metrics in most scenarios IMO.

agreed, except for targeted VOIs.

File: examples/recon_example.py (hunk 3 above)
Comment (casperdcl) [reply examples/recon_example.py]: ah if it's over VOI then sure...

Comment (casperdcl) [reply examples/recon_example.py]:
```suggestion
def MSE(x, y):
    """mean squared error between two arrays"""
    return ((x-y)**2).mean()

def NRMSE(x, y):
    """normalised mean squared error between two arrays"""
    return (((x-y)**2).mean() / (y**2).mean())**0.5
```

File: examples/recon_example.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                                              roi_mask_dict = roi_image_dict,
+                                              metrics_dict = {'MSE':MSE, 
+                                                              'MAE':MAE, 
+                                                              'PSNR':PSNR},
+                                              statistics_dict = {'MEAN': (lambda x: x.mean()),
+                                                                 'STDDEV': (lambda x: x.std()),
+                                                                 'MAX': (lambda x: x.max()),
+                                                                 'COM': (lambda x: np.array([3,2,1]))},
+                                            #   post_smoothing_fwhms_mm_list = [5., 8.]
+                                              )
```
Comment (casperdcl) [inline examples/recon_example.py]:
```suggestion
    roi_mask_dict=roi_image_dict,
    metrics_dict={'NRMSE': NRMSE, 'MSE': MSE, 'MAE': MAE, 'PSNR': PSNR},
    statistics_dict={
        'MEAN': (lambda x: x.mean()), 'STDDEV': (lambda x: x.std()),
        'MAX': (lambda x: x.max()), 'COM': (lambda x: np.array([3,2,1]))},
    # post_smoothing_fwhms_mm_list = [5., 8.]
)
```

File: examples/recon_example.py (hunk 2 above)
Comment (paskino) [inline examples/recon_example.py]:
```suggestion
```

**Review by casperdcl — APPROVED** [review]
fine for now

**Pull request merged**